CMakeLists.txt

```diff
@@ -0,0 +1,276 @@
+cmake_minimum_required(VERSION 3.21)
+
+project(vllm_extensions LANGUAGES CXX)
+
+message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
+
+include(${CMAKE_CURRENT_LIST_DIR}/cmake/utils.cmake)
+
+#
+# Supported python versions.  These versions will be searched in order, the
+# first match will be selected.  These should be kept in sync with setup.py.
+#
+set(PYTHON_SUPPORTED_VERSIONS "3.8" "3.9" "3.10" "3.11")
+
+# Supported NVIDIA architectures.
+set(CUDA_SUPPORTED_ARCHS "7.0;7.5;8.0;8.6;8.9;9.0")
+
+# Supported AMD GPU architectures.
+set(HIP_SUPPORTED_ARCHS "gfx908;gfx90a;gfx942;gfx1100")
+
+#
+# Supported/expected torch versions for CUDA/ROCm.
+#
+# Currently, having an incorrect pytorch version results in a warning
+# rather than an error.
+#
+# Note: the CUDA torch version is derived from pyproject.toml and various
+# requirements.txt files and should be kept consistent.  The ROCm torch
+# versions are derived from Dockerfile.rocm
+#
+set(TORCH_SUPPORTED_VERSION_CUDA "2.1.2")
+set(TORCH_SUPPORTED_VERSION_ROCM_5X "2.0.1")
+set(TORCH_SUPPORTED_VERSION_ROCM_6X "2.1.1")
+
+#
+# Try to find python package with an executable that exactly matches
+# `VLLM_PYTHON_EXECUTABLE` and is one of the supported versions.
+#
+if (VLLM_PYTHON_EXECUTABLE)
+  find_python_from_executable(${VLLM_PYTHON_EXECUTABLE} "${PYTHON_SUPPORTED_VERSIONS}")
+else()
+  message(FATAL_ERROR
+    "Please set VLLM_PYTHON_EXECUTABLE to the path of the desired python version"
+    " before running cmake configure.")
+endif()
+
+#
+# Update cmake's `CMAKE_PREFIX_PATH` with torch location.
+#
+append_cmake_prefix_path("torch" "torch.utils.cmake_prefix_path")
+
+#
+# Import torch cmake configuration.
+# Torch also imports CUDA (and partially HIP) languages with some customizations,
+# so there is no need to do this explicitly with check_language/enable_language,
+# etc.
+#
+find_package(Torch REQUIRED)
+
+# For some reason torch does not add libtorch_python.so to the list of torch
+# libraries to link.  Find it by hand using `append_torchlib_if_found` from
+# torch's cmake setup.
+append_torchlib_if_found(torch_python)
```

Comment:
Have you looked more into this? It is very odd that this is needed :) What are the libraries that are currently included in `${TORCH_LIBRARIES}`? Which symbols are missing from them?

Reply:
The first undefined symbol I encountered was `pybind11::detail::type_caster<at::Tensor, void>::load(pybind11::handle, bool)` but there may be more.

This symbol is defined in `libtorch_python.so` which would normally have been added by `torch.utils.cpp_extension.CUDAExtension` if setup.py was using torch's extension mechanism to build.

The cmake config from `find_pacakge(Torch)` does not add this library since I think it assumes it is only going to be used for standalone C++ binaries that link against torch.   See `torch/share/cmake/Torch/TorchConfig.cmake` where it adds all the other libraries.

Reply:
I'll update the comment to include the details.